README
Config-Format-Ini version 0.01
==============================
 
 This module reads INI files by following the spec presently found at
http://www.cloanto.com/specs/ini.html .  It supports most of the spec,
including multi-valued keys (separated by commas), double-quoted val-
ues, and free comments, and overide of earlier sections and keys.
Escape and continuation strings are in the TODO list.


INSTALLATION

To install this module type the following:

   perl Makefile.PL    ( or perl Build.PL )
   make
   make test
   make install

DEPENDENCIES

 Attribute::Handlers
 Parse::RecDescent
 File::Slurp    


The following Test modules are not required, but are recomended for testing:
Test::Pod     
Test::Pod::Coverage 
Test::HasVersion
Test::CPAN::Meta
Test::Exception
Test::Distribution
Test::Dependencies
Test::MinimumVersion
Test::Strict
Test::Spelling
                    

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2006 by Ioannis Tambouras

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.4 or,
at your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Config::Format::Ini.

0.08  Sat Aug 28 06:06:16 2011
       - Regular upkeep: changed installer, changed and added tests

0.07  Wed Jul 23 21:48:23 2008
        - Fixed license entry (to perl)

0.06  Tue Jul  1 12:20:44 2008
	- Fixt Perl dependacy to 5.008

0.05  Tue Jul  1 12:20:44 2008
	- added global $Config::Format::Init::SIMPLIFY
          When enabled (disabled by default), single-element-lists
	  are converted to simple string; that is,
          the array_ref [ 'john'] is converted to string 'john';
          In addition, if the list (or section) has is blank, it
          results to a simple undef (instead [] in previous versions).

0.04  Sun Oct  1 01:54:02 2006
	- values can occupy multiple lines if continued with \<newline>
	- Added tests for this continuation charter:  '\', followed by "\n" .

0.03  Thu Sep 28 03:21:24 2006
	- values can contain \x000 and ansi escapes
	- Added tests for escaped hexdigits and ansi chars

0.02  Tue Sep 26 06:03:39 2006
	- Unrecognized keys are skipped
	- Unrecognized sections are skipped
	- Added resync tests

0.01  Sun Sep 24 18:03:39 2006
	- original version; created by h2xs 1.23 with options
		-AX -n Config::Format::Ini



Build.PL
use Module::Build;
my $b= Module::Build->new(
        module_name        => 'Config::Format::Ini',
        dist_version_from  => 'lib/Config/Format/Ini.pm',
        dist_author        => 'Ioannis Tambouras <ioannis@cpan.org>',
        requires           => {  perl                => '5.008',
                                'Attribute::Handlers'=>  0 ,
                                'Parse::RecDescent'  =>  0 ,
                                'File::Slurp'        =>  0 ,
                              },
        recommends         => { 'Test::More'         => 0,
                                'Test::Exception'    => 0,
                                'Test::Pod::Coverage'=> '1.0',
                                'Test::Pod'          => 0,
								'Test::HasVersion'    => '0.012',
								'Test::CPAN::Meta'    => '0.18',
								'Test::Exception'     => '0.31',
								'Test::Distribution'  => '2.00',
								'Test::Dependencies'  => '0.12',
								'Test::MinimumVersion'=> '0.101080',
								'Test::Strict'        => '0.14',
								'Test::Spelling'      => '0.15',
                              },
        license            => 'perl',
        #create_makefile_pl=> 'traditional',
        #create_readme     => 1,
);

$b->create_build_script;


lib/Config/Format/Ini.pm
package Config::Format::Ini;

use 5.008000;
use strict;
use warnings;

use base qw( Exporter );
our @EXPORT = qw( read_ini  );
our %EXPORT_TAGS = ( 'all' => [ qw( ) ] );
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our $VERSION = '0.08';

use Config::Format::Ini::Grammar;
use File::Slurp qw(slurp);

our $SIMPLIFY=0;

sub read_ini {
	return unless @_;
	my $msg;
        $msg  .= scalar slurp $_  for @_ ;
	my $p  = new Config::Format::Ini::Grammar;
	my $result = $p->startrule( $msg );
	_simplify( $result)  if $SIMPLIFY;
	$result;
}

sub _arr2scalar {
        # change arrays with one element to string
        my $ref = shift ||return;
        return unless ref $ref eq 'HASH';
        while (my( $k,$v )=each %$ref) {
                next unless ref $v eq 'ARRAY';
                (1 == @$v ) and $ref->{$k} = $v->[0];
                (0 == @$v ) and $ref->{$k} = undef;
        }
}
sub _simplify {
        my $ini =shift;
        return unless ref $ini eq 'HASH';
        (0 == keys %{$ini->{$_}})
                ? undef $ini->{$_}
                : _arr2scalar  $ini->{$_}
                for (keys %$ini);
}


1;
__END__

=head1 NAME

Config::Format::Ini - Reads INI configuration files

=head1 SYNOPSIS

  use Config::Format::Ini;

=head1 DESCRIPTION

This module reads INI files by following the spec
presently found at http://www.cloanto.com/specs/ini.html .
It supports most of the spec, including multi-valued keys (separated by
commas), double-quoted values, and free comments, and override of earlier
sections and keys.

Escape and continuation strings are in the TODO list.

=head2 EXPORT

=over

=item read_ini() 
Reads ini data from filename, or filenames when multiple 
arguments are supplied.


=back


=head1 SEE ALSO

Config::Ini, Config::INI::Simple, Config::IniFiles, and
OpenInteract2::Config::Ini .  There is a paper about the INI format
at http://www.cloanto.com/specs/ini.html .

=head1 AUTHOR

Ioannis Tambouras, E<lt>ioannis@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Ioannis Tambouras

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.


=cut


lib/Config/Format/Ini/Grammar.pm
package  Config::Format::Ini::Grammar;
use Parse::RecDescent;
use strict;
use warnings;
#use 5.008000;
our $VERSION='0.01';

my $gram = <<'END_GRAMMAR' ;
{       my $h ;
	use Data::Dumper;
	my %esc =  map {$_ => pack 'H2', $_}  0..255;
	@esc{ '\n', '\t', '\r', '\a' } = ( "\n", "\t", "\r", "\a" );
	sub merge2hash { my $aref = shift; my $h;
	   ref eq 'HASH' and @{$h}{keys %$_} = values %$_  for @$aref; $h;
	}
        sub  postprocess {
		local $_ = shift ||return;
		s/ \\x([\d]{1,3})/ $esc{ "$1"+0 } /egx;
		s/ (\\[ntra])    / $esc{$1}       /xge;
	        $_ ;
        }
}
	         

    startrule : <skip:'[ \t]*'>   
		Section(s) 
		{ $return =  merge2hash( $item[2])  }

    Section:   Title  "\n"  Pair(s?) 
		{ $return = { $item[1]=> merge2hash( $item[3])||{} } }
                | COMMENT(s) 
                | BLANK(s) 
  	        | <resync>

    Title:     '[' /\w+/ ']'      COM(?)  
		{ $item[2] }

    Pair:      KEY '=' VAL(s?)    COM(?)   "\n"
	       {$return = {$item[1]=>$item[3]} }
    Pair:      COMMENT(s){$return = 0}
              | BLANK(s) {$return = 0}
    Pair:     <resync:[^[\n]+\n>

    KEY:      /\w+(?=\s*=)/

    VAL:       <perl_quotelike>
               <reject: ${item[1]}[1] ne '"' >
               { $item[1]->[2] =~ s/\\\n//g }
               { $item[1]->[2] }
    VAL:      
                /.*? (?<!\\)(?=[,;#\n]) /sx 
                { $item[1] =~ s/\\\n//g     }
		{ $item[1] =~ s/[;#].*\Z//m }
		{ $item[1] =~ s/^\s+|\s+$// }
                { postprocess( $item[1] ) }
    VAL:       ',' VAL 
               {$item[2]} 
 


    BLANK:     "\n"
    COM:        /^(?:[#;].*)/
    COMMENT:    COM "\n"

END_GRAMMAR

sub new { Parse::RecDescent->new( $gram ) }

1;
__END__
=head1 NAME

Config::Format::Ini::Grammar -  P:RD grammar for ini file format



Makefile.PL
use 5.008008;
use inc::Module::Install;
name           'Config-Format-Ini'              ;
license        'perl'                           ;
all_from       'lib/Config/Format/Ini.pm'       ;
abstract_from  'lib/Config/Format/Ini.pm'       ;
no_index       'directory'   => 'examples'      ;
author         'Ioannis Tambouras <ioannis@cpan.org>';

requires
#     'Getopt::Compact'     => '0.04',
#	     'List::MoreUtils'     => '0.32',
#         'namespace::autoclean'=> '0.12',
#	     'List::Util'          => '1.23',
         'Attribute::Handlers' => 0,
	     'Parse::RecDescent'   => 0,
	     'File::Slurp'         => '9999.18',
;
test_requires 
              'Test::Pod'           => '1.45',
              'Test::Pod::Coverage' => '1.08',
              'Test::HasVersion'    => '0.012',
              'Test::CPAN::Meta'    => '0.18',
              'Test::Exception'     => '0.31',
              'Test::Distribution'  => '2.00';
              'Test::Dependencies'  => '0.12',
              'Test::MinimumVersion'=> '0.101080',
              'Test::Strict'        => '0.14', 
              'Test::Spelling'      => '0.15',
;
auto_install;
WriteAll;
